diff --git a/dwm.c b/dwm.c
index 0362114..c313b5e 100644
--- a/dwm.c
+++ b/dwm.c
@@ -233,6 +233,11 @@ static int xerror(Display *dpy, XErrorEvent *ee);
 static int xerrordummy(Display *dpy, XErrorEvent *ee);
 static int xerrorstart(Display *dpy, XErrorEvent *ee);
 static void zoom(const Arg *arg);
+static void bstack(Monitor *m);
+static void bstackhoriz(Monitor *m);
+static void centeredmaster(Monitor *m);
+static void centeredfloatingmaster(Monitor *m);
+static void deck(Monitor *m);

 /* variables */
 static const char broken[] = "broken";
